game/src/layer/problems.rs

```diff
@@ -0,0 +1,191 @@
+            g.redraw(&self.draw);
+        }
+    }
+    fn draw_minimap(&self, g: &mut GfxCtx) {
+        g.redraw(&self.draw);
+    }
+}
+
+impl ProblemMap {
+    pub fn new(ctx: &mut EventCtx, app: &App, opts: Options) -> ProblemMap {
+        let mut pts = Vec::new();
+        for (trip, problems) in &app.primary.sim.get_analytics().problems_per_trip {
+            for problem in problems {
+                if opts.show(app.primary.sim.trip_info(*trip), problem) {
+                    pts.push(match problem {
+                        Problem::IntersectionDelay(i, _)
+                        | Problem::LargeIntersectionCrossing(i) => {
+                            app.primary.map.get_i(*i).polygon.center()
+                        }
+                        Problem::OvertakeDesired(on) => on.get_polyline(&app.primary.map).middle(),
+                    });
+                }
+            }
+        }
+
+        let mut batch = GeomBatch::new();
+        let legend = if let Some(ref o) = opts.heatmap {
+            Some(make_heatmap(
+                ctx,
+                &mut batch,
+                app.primary.map.get_bounds(),
+                pts,
+                o,
+            ))
+        } else {
+            let circle = Circle::new(Pt2D::new(0.0, 0.0), Distance::meters(10.0)).to_polygon();
+            // TODO Different colors per problem type
+            for pt in pts {
+                batch.push(Color::RED.alpha(0.8), circle.translate(pt.x(), pt.y()));
+            }
+            None
+        };
+        let controls = make_controls(ctx, app, &opts, legend);
+        ProblemMap {
+            time: app.primary.sim.time(),
+            opts,
+            draw: ctx.upload(batch),
+            panel: controls,
+        }
+    }
+
+    fn options(&self) -> Options {
+        let heatmap = if self.panel.is_checked("Show heatmap") {
+            Some(HeatmapOptions::from_controls(&self.panel))
+        } else {
+            None
+        };
+        let mut modes = BTreeSet::new();
+        for m in TripMode::all() {
+            if self.panel.is_checked(m.ongoing_verb()) {
+                modes.insert(m);
+            }
+        }
+        Options {
+            heatmap,
+            modes,
+            show_delays: self.panel.is_checked("show delays"),
+            show_large_crossings: self
+                .panel
+                .is_checked("show crossings over large intersections"),
+            show_overtakes: self
+                .panel
+                .is_checked("show where cars want to overtake cyclists"),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Options {
+    // If None, just a dot map
+    heatmap: Option<HeatmapOptions>,
+    modes: BTreeSet<TripMode>,
+    show_delays: bool,
+    show_large_crossings: bool,
+    show_overtakes: bool,
+    // TODO Time range
+}
+
+impl Options {
+    pub fn new() -> Options {
+        Options {
+            heatmap: None,
```

Comment:
I thought the heatmap would be slow to calculate, but actually it's not, and it seems easier to spot major areas faster. Opinions?

Reply:
Seems reasonable to me. 👍